pkg/rtc/participant.go

```diff
@@ -4107,3 +4133,125 @@ func (p *ParticipantImpl) AddTransceiverFromTrackLocal(
 		)
 	}
 }
+
+func (p *ParticipantImpl) handleIncomingRpcAck(requestId string) bool {
+	p.rpcLock.Lock()
+	defer p.rpcLock.Unlock()
+
+	handler, ok := p.rpcPendingAcks[requestId]
+	if !ok {
+		return false
+	}
+
+	handler.Resolve()
+	delete(p.rpcPendingAcks, requestId)
+	return true
+}
+
+func (p *ParticipantImpl) handleIncomingRpcResponse(requestId string, payload string, err *sutils.RpcError) bool {
+	p.rpcLock.Lock()
+	defer p.rpcLock.Unlock()
+
+	handler, ok := p.rpcPendingResponses[requestId]
+	if !ok {
+		return false
+	}
+
+	handler.Resolve(payload, err)
+	delete(p.rpcPendingResponses, requestId)
+	return true
+}
+
+func (p *ParticipantImpl) PerformRpc(req *livekit.PerformRpcRequest, resultCh chan string, errorCh chan error) {
+	responseTimeout := req.GetResponseTimeoutMs()
+	if responseTimeout <= 0 {
+		responseTimeout = 10000
+	}
+	responseTimeoutWithoutRTT := max(0, responseTimeout-uint32(sutils.RpcMaxRoundTripLatency))
+
+	go func() {
+		if len([]byte(req.GetPayload())) > sutils.RpcMaxPayloadBytes {
+			errorCh <- &sutils.RpcError{Code: sutils.RpcRequestPayloadTooLarge}
+			return
+		}
+
+		id := uuid.New().String()
+
+		responseTimer := time.AfterFunc(time.Duration(responseTimeout)*time.Millisecond, func() {
+			p.rpcLock.Lock()
+			delete(p.rpcPendingResponses, id)
+			p.rpcLock.Unlock()
+
+			select {
+			case errorCh <- &sutils.RpcError{Code: sutils.RpcResponseTimeout}:
+			default:
+			}
+		})
+		ackTimer := time.AfterFunc(sutils.RpcMaxRoundTripLatency, func() {
+			p.rpcLock.Lock()
+			delete(p.rpcPendingAcks, id)
+			delete(p.rpcPendingResponses, id)
+			p.rpcLock.Unlock()
+			responseTimer.Stop()
+
+			select {
+			case errorCh <- &sutils.RpcError{Code: sutils.RpcConnectionTimeout}:
+			default:
+			}
+		})
+
+		rpcRequest := &livekit.DataPacket{
+			Kind: livekit.DataPacket_RELIABLE,
+			Value: &livekit.DataPacket_RpcRequest{
+				RpcRequest: &livekit.RpcRequest{
+					Id:                id,
+					Method:            req.GetMethod(),
+					Payload:           req.GetPayload(),
+					ResponseTimeoutMs: responseTimeoutWithoutRTT,
+					Version:           1,
+				},
+			},
+		}
+		data, err := proto.Marshal(rpcRequest)
+		if err != nil {
+			ackTimer.Stop()
```

Comment:
make a little local function to do this and call it, looks like a couple of places is doing this, so just to make sure there are no timer leaks etc, make a function and use it, if later another error case pops up, it can be re-used.

Reply:
Had the local function change but reverted it because it was making the code harder to read because of the lack of named params in Go. 

It's only 2-3 places, I think we can keep it as is.